### PR TITLE
Spectrum m4i improvements

### PR DESCRIPTION
**Pull request by sldesnoo-Delft**
Added asynchronous acquisition with setup_multi_recording(), start_triggered() and get_data()
Add 20 MHz low-pass filter
Added firmware to get_idn()
Improved error logging
Use f-strings
Improved documentation

### CODE REVIEW

**Comment by FarBo** [comment]
@sldesnoo-Delft 
Thanks for the PR. I will review it as soon as possible.

File: qcodes_contrib_drivers/drivers/Spectrum/M4i.py [hunk 1]
```diff
@@ -32,7 +32,8 @@
         log.info('M4i: adding header_dir %s to sys.path' % header_dir)
         sys.path.append(header_dir)
     import pyspcm
-except (ImportError, OSError) as ex:
+    import py_header.spcerr
```
Comment (FarBo) [inline qcodes_contrib_drivers/drivers/Spectrum/M4i.py:35]: Is there any specific reason to import this here?

File: qcodes_contrib_drivers/drivers/Spectrum/M4i.py [hunk 2]
```diff
@@ -695,59 +733,152 @@ def _read_channel(self, channel, memsize=None):
...
+            self.set(f'anti_aliasing_filter_{channel_index}', lp_filter)
+        self.set(f'range_channel_{channel_index}', mV_range)  # note: set after voltage range
         # can only be used with DC coupling and 50 Ohm path (hf)
         if compensation is not None:
-            getattr(self, 'ACDC_offs_compensation_{}'.format(
-                channel_index))(compensation)
+            self.set(f'ACDC_offs_compensation_{channel_index}', compensation)
 
     def set_ext0_OR_trigger_settings(self, trig_mode, termination, coupling, level0, level1=None):
+        ''' Configures ext0 trigger
```
Comment (FarBo) [inline qcodes_contrib_drivers/drivers/Spectrum/M4i.py:769]: Double quote for docstring is prefered.

File: qcodes_contrib_drivers/drivers/Spectrum/M4i.py [hunk 3]
```diff
@@ -695,59 +733,152 @@ def _read_channel(self, channel, memsize=None):
...
+            n_triggers (int): total number of triggers
+            pretrigger_size (int): size of data trace before triggering
+            boxcar_average (bool): use mode 'boxcar average'
+
+        Example:
+            digitizer.setup_multi_recording(size, n_triggers)
+            digitizer.start_triggered()
+            data = digitizer.get_data()
+            digitizer.start_triggered()
+            data = digitizer.get_data()
```
Comment (FarBo) [inline qcodes_contrib_drivers/drivers/Spectrum/M4i.py:820]: Not sure why digitizer.start_triggered() and data = digitizer.get_data() are repeated here?

File: qcodes_contrib_drivers/drivers/Spectrum/M4i.py [hunk 4]
```diff
@@ -695,59 +733,152 @@ def _read_channel(self, channel, memsize=None):
...
+            pretrigger_size = self.pretrigger_memory_size()
+        posttrigger_size = self._hw_memsize(posttrigger_size)
+        seg_size = self._hw_memsize(posttrigger_size+pretrigger_size)
+        memsize = self._hw_memsize(n_triggers * seg_size)
+        self.data_memory_size(memsize)
+        self.segment_size(seg_size)
+        self.posttrigger_memory_size(posttrigger_size)
+
+    def start_triggered(self):
+        '''Starts triggered acquisition
```
Comment (FarBo) [inline qcodes_contrib_drivers/drivers/Spectrum/M4i.py:841]: double quote

File: qcodes_contrib_drivers/drivers/Spectrum/M4i.py [hunk 5]
```diff
@@ -695,59 +733,152 @@ def _read_channel(self, channel, memsize=None):
...
+        '''
+        self.general_command(pyspcm.M2CMD_CARD_START
+                             | pyspcm.M2CMD_CARD_ENABLETRIGGER)
+
+    def get_data(self):
+        ''' Reads measurement data from the digitizer.
```
Comment (FarBo) [inline qcodes_contrib_drivers/drivers/Spectrum/M4i.py:847]: double quote

**Review by FarBo — APPROVED** [review]

File: qcodes_contrib_drivers/drivers/Spectrum/M4i.py (hunk 5 above)
Comment (sldesnoo-Delft) [reply qcodes_contrib_drivers/drivers/Spectrum/M4i.py:847]: @FarBo, Hi Farzad, you commented on the code, but also approved the changes. Do I still have to fix the code?

**Review by FarBo — APPROVED** [review]
@sldesnoo-Delft 
The PR is fine. I have a few minor comments. It would be nice to close these comments before the merge.

File: qcodes_contrib_drivers/drivers/Spectrum/M4i.py (hunk 5 above)
Comment (FarBo) [reply qcodes_contrib_drivers/drivers/Spectrum/M4i.py:847]: I forgot to write a final comment in the approve. Now, it is there :)

File: qcodes_contrib_drivers/drivers/Spectrum/M4i.py (hunk 1 above)
Comment (sldesnoo-Delft) [reply qcodes_contrib_drivers/drivers/Spectrum/M4i.py:35]: spcerr is needed and it is in the directory added to the path on line 33. So, yes, the import is needed.

**Pull request merged**